boot/bootutil/include/bootutil/boot_status.h

```diff
@@ -0,0 +1,150 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __BOOT_STATUS_H__
+#define __BOOT_STATUS_H__
+
+#include <stdint.h>
+#include <stddef.h>
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * The shared data between boot loader and runtime SW is TLV encoded. The
+ * shared data is stored in a well known location in memory and this is a
+ * contract between boot loader and runtime SW.
+ *
+ * The structure of shared data must be the following:
+ *  - At the beginning there must be a header: struct shared_data_tlv_header
+ *    This contains a magic number and a size field which covers the entire
+ *    size of the shared data area including this header.
+ *  - After the header there come the entries which are composed from an entry
+ *    header structure: struct shared_data_tlv_entry and the data. In the entry
+ *    header is a type field (tly_type) which identify the consumer of the
+ *    entry in the runtime SW and specify the subtype of that data item. There
+ *    is a size field (tlv_len) which covers the size of the the data. After
+ *    this structure comes the actual data.
+ *
+ *  - Arbitrary number and size of data entry can be in the shared memory area.
+ *
+ * This table gives of overview about the tlv_type field in the entry header.
+ * The tlv_type always composed from a major and minor number. Major number
+ * identifies the addressee in runtime SW, who should process the data entry.
+ * Minor number used to encode more info about the data entry. The actual
+ * definition of minor number could change per major number.
+ *
+ * In case of boot status data, which can be processed by an attestation
+ * service the minor number is split further to two part: sw_module and claim.
+ * The sw_module identifies the SW component in the system which the data item
+ * belongs to and the claim part identifies the exact type of the data.
+ *
+ * |---------------------------------------|
+ * |            tlv_type (16)              |
+ * |---------------------------------------|
+ * | tlv_major(4)|      tlv_minor(12)      |
+ * |---------------------------------------|
+ * | MAJOR_IAS   | sw_module(6) | claim(6) |
+ * |---------------------------------------|
+ */
+
+/* General macros to handle TLV type */
+#define MAJOR_MASK 0xF     /* 4  bit */
+#define MAJOR_POS  12      /* 12 bit */
+#define MINOR_MASK 0xFFF   /* 12 bit */
+
+#define SET_TLV_TYPE(major, minor) \
+        (((uint16_t)((major) & MAJOR_MASK) << MAJOR_POS) \
+        | ((minor) & MINOR_MASK))
+#define GET_MAJOR(tlv_type) ((uint16_t)(tlv_type) >> MAJOR_POS)
+#define GET_MINOR(tlv_type) ((tlv_type) & MINOR_MASK)
+
+/* Magic value which marks the beginning of shared data area in memory */
```

Comment:
This seems like a fairly weak protection of this shared data (imagine uninitialized memory). I wonder if we shouldn't either have a longer magic value, or hash the magic value, and verify that the hash is correct?

Reply:
This TLV is landing in shared memory. If only secure domain has write access here - then non-secure one could trusted that if magic value is right, the TLV is OK.
@davidvincze  Am I right.

Reply:
I'm concerned about uninitialized memory having the magic value. If it is random, then 1:65536 devices with spurious data seems rather high.

Reply:
Yes, @nvlsianpu  is right.
That 1:65536 ratio really is troublesome. I think the only reason to check whether this area is already initialized is the fact that there can be multiple images. I think there should be an extra function call before calling `boot_save_boot_status()` function to initialize the shared data area. It means no one else can use this area before mcuboot. It sound acceptable to me (at 8:00 PM).

Reply:
Hashing the data seems a bit overkill. Basically we do not want to protect against malicious modification. Shared memory must be only secure accessible. Hashing would require to have the same hashing capability in the runtime firmware, which might not the case. I  think the best would be to increase the size of the magic value. @d3zd3z What do you think how many bits would provide enough protection? 128?

Reply:
128 is a bit excessive, 64 bits should be sufficient for a magic value. 32 might even be enough.
It is a little better, because this value will only matter in a configuration where somehow someone configures an mcuboot without shared data and a TFM that has it. It shouldn't happen, but might.

Reply:
32 bit seems reasonable, because as you wrote miss configuration also needs to happen. TF-M and MCUBoot has the same config option to turn on/off shared data support at compile time. So issue could happen if miss-configuration happens. i.e.: TF-M expects shared data, but bootloader does not provide it and additionally the random data in the shared memory equals with the magic number. TF-M v1.0 release is very close, so if magic value will be changed then this cause an incompatibility. It can be handled in later TF-M version, but might better to avoid. @d3zd3z: Do you have any idea to avoid?